src/charm.py

```diff
@@ -2,203 +2,144 @@
+        self._container_name = "spark"
+        self.container = self.unit.get_container(self._container_name)
+
+        self.framework.observe(self.on.spark_pebble_ready, self._on_spark_pebble_ready)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.remove, self._on_remove)
+
+    @property
+    def _template_files(self):
+        src_dir = Path("src")
+        manifests = [file for file in glob.glob(f"{src_dir}/*.yaml")]
+        return manifests
+
+    @property
+    def _context(self):
+        context = {
+            "app_name": self.model.app.name,
+            "model_name": self.model.name,
+        }
+        return context
+
+    @property
+    def _pebble_layer(self):
+        pebble_layer = {
+            "summary": "spark layer",
+            "description": "pebble config layer for spark-k8s",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Spark Operator layer",
+                    "startup": "enabled",
+                    "command": (
+                        f"/usr/bin/tini -s -- /usr/bin/spark-operator -v=2 "
+                        "-logtostderr "
+                        f"-namespace={self.model.name} "
+                        "-enable-ui-service=true "
+                        "-controller-threads=10 "
+                        "-resync-interval=30 "
+                        "-enable-batch-scheduler=false "
+                        "-enable-metrics=true "
+                        "-metrics-labels=app_type "
+                        f"-metrics-port={self.model.config['metrics-port']} "
+                        "-metrics-endpoint=/metrics "
+                        "-enable-resource-quota-enforcement=false "
+                        "-enable-webhook=true "
+                        f"-webhook-svc-namespace={self.model.name} "
+                        f"-webhook-port={self.model.config['webhook-port']} "
+                        f"-webhook-svc-name={self.model.app.name} "
+                        f"-webhook-config-name={self._mutating_webhook_name} "
+                        f"-webhook-namespace-selector=model.juju.is/name={self.model.name} "
+                        "-webhook-fail-on-error=true"
+                    ),
                 }
             },
+        }
+        return pebble_layer
+
+    def _on_spark_pebble_ready(self, _):
+        self.unit.status = MaintenanceStatus("Configuring Spark Charm")
+
+        self.container.push("/etc/webhook-certs/ca-cert.pem", self._stored.ca, make_dirs=True)
+        self.container.push(
+            "/etc/webhook-certs/server-cert.pem", self._stored.cert, make_dirs=True
         )
-        self.model.unit.status = ActiveStatus()
+        self.container.push("/etc/webhook-certs/server-key.pem", self._stored.key, make_dirs=True)
+
+        self.resource_handler.apply()
```

Comment:
What's the reason for placing the `apply` here?
This `apply` feels weird in the pebbleReady event handler, I would dedicate a helper method for just applying on install and config changed, that way we can guarantee the resources are created at the beginning and changed/patched if needed after a config file.

Reply:
I think of the pebble ready event as the setup event. That's why I am applying the manifests here. What should be in the pebble ready event handler, only pebble related actions?  I could move it into a separate function. It won't observe config changed though cause the manifests won't change.

Reply:
`<charm>_pebble_ready` is used more for configuring how services should be launched. Look [here](https://operator-framework.readthedocs.io/en/latest/index.html?highlight=install%20event#ops.charm.PebbleReadyEvent) for more info. It is also important to notice that pebble ready events can be triggered "any time whatsoever during the lifecycle of a charm;", thus applying your manifests here could lead to an undesired behaviour.